h2o-jetty-9/src/main/java/water/webserver/jetty9/Jetty9WebsocketServlet.java

```diff
@@ -68,9 +67,30 @@ public void onWebSocketError(Throwable cause) {
         }
 
     }
-    
+
+    /**
+     * Please note, each Servlet has it's own instance of WebSocketServletFactory.
+     *
+     * @param factory Factory object to register socket creator with.
+     */
     @Override
     public void configure(WebSocketServletFactory factory) {
-        factory.register(Jetty9WebsocketHandler.class);
+        factory.setCreator(new H2OWebSocketCreator());
+    }
+
+    /**
+     * Custom in-place socket creator, returning new instance of {@link Jetty9WebsocketHandler}m
+     * which already contains the proper {@link WebsocketServlet} implementation the request is being delegated to.
+     * <p>
+     * This is required, as default {@link WebSocketServletFactory} uses {@link org.eclipse.jetty.util.DecoratedObjectFactory}
+     * to instantiate {@link WebSocketListener} classes. This class is only able to instantiate static classes with 0-arg constructor,
+     * which inner non-static class {@link Jetty9WebsocketHandler} is NOT.
+     */
+    public class H2OWebSocketCreator implements WebSocketCreator {
```

Comment:
@honzasterba Please review this part. There is a problem with Jetty 9 - it's unable to instantiate `Jetty9WebsocketHandler` which delegates the calls to the proper implementation. It's all described for everyone in the Javadoc in detail - the default factory is only able to instantiate static classes with 0-arg constructors, our class inner non-static. As mentioned in the JavaDoc, each servlet has it's own factory instance - this means the custom `H2OWebSocketCreator` is only assigned to a servlet-local factory, not a global one.

Reply:
this looks like the best solution
knit: H2OWebSocketCreator does not need to be public
self-shaming: I wish I named `water.webserver.iface.WebsocketServlet` differently (with H2O prefix), now it is confusing with jetty's `WebSocketServlet`

Reply:
Thanks, I'll remove the public keyword once the tests which are in the middle of finishing actually finish. 

I guess this is internal and we can always rename it btw :wink:

Reply:
Good call, will do that after you merge and before we release